### PR TITLE
kubeless 0.3.0 (new formula)

### PR DESCRIPTION
**Pull request by andresmgot**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/kubeless.rb
Comment (andresmgot) [inline Formula/kubeless.rb]: Note that we cannot test much else since we need a Kubernetes environment configured locally in the machine

**Review by apjanke — DISMISSED** [review]
This looks like it is installing from a binary distribution. Generally we require formula in Homebrew core to be built from source. Could this be modified to download and build from the source distribution?

**Comment by andresmgot** [comment]
Updated build from source in the formula

**Comment by andresmgot** [comment]
I had to add `go` as a dependency, I am not sure if adding a ~300MB package worths just for building kubeless from source (45MB), is that expected @apjanke?

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: This line is no longer true.

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: What happens if we don't pass the `commit`?

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: Wrap this to <=80 characters.

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: Kinda feel like this should just be using a `depends_on "kubernetes-cli" => :recommended` dep.

**Comment by DomT4** [comment]
> is that expected

Yes. As long as users pour the bottle they'll never see the `go` dependency here.

File: Formula/kubeless.rb
Comment (andresmgot) [reply Formula/kubeless.rb]: when running `kubeless version` it gives you the information about the version (release tag) and the commit of that build. If not given it sets the commit to `HEAD` and that output will differ from the output that the `kubeless` binary we distribute gives.

Comment (andresmgot) [reply Formula/kubeless.rb]: I will move that to a property and cover it with the test

File: Formula/kubeless.rb
Comment (andresmgot) [reply Formula/kubeless.rb]: True, the `kubernetes-cli` is recommended but what the caveat try to say is that a cluster should be ready and how to install the server side of Kubeless

**Comment by andresmgot** [comment]
Thanks for the review. _Updating..._

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: Let's just use the `git` checkout if it absolutely needs a `commit` rather than starting to do things like this, which absolutely nobody new to Ruby is going to understand.

File: Formula/kubeless.rb
Comment (DomT4) [reply Formula/kubeless.rb]: Generally Homebrew leaves basic usage instructions to the upstream READMEs to handle, not least because people barely read `caveats`. I'll let ILZ or another maintainer chime in on that though.

File: Formula/kubeless.rb
Comment (andresmgot) [reply Formula/kubeless.rb]: okay, I was unaware of that option

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: Should be an empty line between the URL blob and the first `depends_on`.

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: @ilovezfs What style of indent are we using here these days? 🙈

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: Use `Utils.popen_read("git rev-parse --short HEAD").chomp` instead. It's better for my long-term happiness 🤣.

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: Use [this sort of format for ldflags](https://github.com/Homebrew/homebrew-core/blob/6d2f9f6fd82923814ac99f56aed674a38dfa4ff7/Formula/ssh-vault.rb#L25-L26). Using `\` in formulae to split lines is just... grim.

File: Formula/kubeless.rb
Comment (andresmgot) [reply Formula/kubeless.rb]: two spaces is the usual indent increment, I can make it 4 spaces to match with the previous line that seems pretty common as well

Comment (DomT4) [reply Formula/kubeless.rb]: Yeah, my feeling is that we've been using 4 on URLs specifically because it lines up neatly, but I'm not sure how consistently that's being enforced or whether it's being enforced at all really.

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: It's not a huge thing but it might be worth using a `%W[]` style array here. It'll let you drop the commas and quotes and it just looks a bit neater.

**Comment by andresmgot** [comment]
ping @DomT4 @apjanke @ilovezfs, any updates here?

**Comment by DomT4** [comment]
`<<-EOS.undent` needs changing to `<<~EOS` now. Everything else is up to @ilovezfs; I'm sure he has opinions on the caveats.

File: Formula/kubeless.rb
Comment (ilovezfs) [reply Formula/kubeless.rb]: The caveats should be removed.

File: Formula/kubeless.rb
Comment (ilovezfs) [reply Formula/kubeless.rb]: This needs a real test of some kind. See faas-cli for how to mock a server.

File: Formula/kubeless.rb
Comment (andresmgot) [reply Formula/kubeless.rb]: okay, I will remove them then

**Comment by andresmgot** [comment]
Thank you for the revisions @DomT4 @ilovezfs. I removed the caveats and added an additional test. Let me know if anything else is needed.

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by andresmgot** [comment]
Hi @ilovezfs,
yesterday we released the version 0.2.4 of kubeless: should I update this PR or should I wait for it to be merged and send another PR updating it?
Thanks

**Comment by ilovezfs** [comment]
just update this one and force push :)

**Comment by andresmgot** [comment]
@ilovezfs is there anything pending in the merging process?

**Comment by andresmgot** [comment]
sorry @ilovezfs, any news here?

**Comment by andresmgot** [comment]
I updated the formula to a newer version and resolved a weird conflict that appeared. Let me know if I should perform any additional step.

**Comment by andresmgot** [comment]
Mm, not sure why the integration failed since the conflict has been resolved...

**Comment by ilovezfs** [comment]
@andresmgot I fixed it for you.

**Comment by andresmgot** [comment]
@ilovezfs awesome, thank you

File: Formula/kubeless.rb
Comment (ilovezfs) [inline Formula/kubeless.rb]: Following this, I get
```
bash-3.2$ kubeless function ls test
FATA[0000] Error listing function test: couldn't get version/kind; json parse error: invalid character 'O' looking for beginning of value 
```

File: Formula/kubeless.rb
Comment (ilovezfs) [inline Formula/kubeless.rb]: redundant

File: Formula/kubeless.rb
Comment (ilovezfs) [inline Formula/kubeless.rb]: not needed. we already have the other test above

File: Formula/kubeless.rb
Comment (ilovezfs) [inline Formula/kubeless.rb]: (buildpath/"src/github.com/kubeless/kubeless").install buildpath.children

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: Don't really need a block form here for a single line.
```ruby
(testpath/"test.py").write "function_code"
```
will work fine.

File: Formula/kubeless.rb
Comment (andresmgot) [reply Formula/kubeless.rb]: that's because the mock server is returning a `200 OK` with an empty body when it is supposed to return an object with the function information

**Comment by andresmgot** [comment]
@DomT4 @ilovezfs I updated the recipe with your comments but something wrong seems to have happened with the integration:
```
Warning: Homebrew/homebrew-core is not on the master branch
```
any clues?

**Comment by DomT4** [comment]
Might have been a CI hiccup. Try rebasing & force pushing and seeing if it reproduces.

**Comment by andresmgot** [comment]
I would like to merge this is possible. Anything else needed @DomT4 @ilovezfs ?

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: Just to check, does this require Python 2.7+ at runtime or can you pass in older versions without issue? Python 2.7 wasn't added to macOS until after Mac OS X Snow Leopard.

Comment (andresmgot) [reply Formula/kubeless.rb]: It doesn't require Python in the host . The runtime flag `python2.7` is used to deploy a Kubernetes Pod (Docker container) with that language available.

File: Formula/kubeless.rb
Comment (DomT4) [inline Formula/kubeless.rb]: Could this not simply be `"cmd/kubeless"` inside this directory, or?

File: Formula/kubeless.rb
Comment (DomT4) [reply Formula/kubeless.rb]: Cool, thanks for the clarification. I've asked another question above but I think that's the last one I can see. ILZ may well have more, but punting to him on that.

File: Formula/kubeless.rb
Comment (andresmgot) [reply Formula/kubeless.rb]: The goal is to install the binary in the `bin/` directory, isn't it? We could set the output of this command to any directory but then we would need to move it to its final destination.

Comment (DomT4) [reply Formula/kubeless.rb]: Yeah. What I mean is since we're already in `src/github.com/kubeless/kubeless` here why do we need to feed `go build` the full `github.com/kubeless/kubeless/cmd/kubeless` path instead of just `cmd/kubeless`?

Comment (DomT4) [reply Formula/kubeless.rb]: Or `./cmd/kubeless` perhaps, I can't remember off the top of my head which `go` insists is correct.

Comment (andresmgot) [reply Formula/kubeless.rb]: ah, that's because the path should be relative to the GOPATH, inside the `src` directory. From `$GOPATH/src/` the correct relative path is that one. Ref: https://golang.org/doc/code.html#ImportPaths

Comment (DomT4) [reply Formula/kubeless.rb]: So `./cmd/kubeless` should work, no? 🙈

Comment (DomT4) [reply Formula/kubeless.rb]: Yeah, it does. My knowledge of `go` isn't completely shot, reassuringly, heh.

```diff
diff --git a/Formula/kubeless.rb b/Formula/kubeless.rb
index 852ddce8e1..018851a931 100644
--- a/Formula/kubeless.rb
+++ b/Formula/kubeless.rb
@@ -18,7 +18,7 @@ class Kubeless < Formula
         -X github.com/kubeless/kubeless/version.GITCOMMIT=#{commit}
       ]
       system "go", "build", "-o", bin/"kubeless", "-ldflags", ldflags.join(" "),
-             "github.com/kubeless/kubeless/cmd/kubeless"
+             "./cmd/kubeless"
       prefix.install_metafiles
     end
   end
```

```
==> Summary
🍺  /usr/local/Cellar/kubeless/0.3.0: 5 files, 27.9MB, built in 41 seconds
```

Comment (andresmgot) [reply Formula/kubeless.rb]: True, that indeed works, I am just used to use the import path rather than current relative path (to make it work in any folder). In any case if that is a convention in this repo I am fine changing it.

Comment (DomT4) [reply Formula/kubeless.rb]: Makes sense to use the relative path here IMO, since Homebrew should always be building inside the `src/github.com/kubeless/kubeless` folder. Relative paths are the norm here when building from inside a directory, at least.

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
Thanks for the :new: formula and your first contribution to Homebrew, @andresmgot!

**Pull request closed without merging**
